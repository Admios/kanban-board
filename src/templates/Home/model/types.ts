export interface Todo {
  id: string;
  stateId: string;
  text: string;
  position: number;
  owner: string;
}

export interface State {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
