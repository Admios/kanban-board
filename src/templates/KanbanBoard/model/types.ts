export interface Task {
  id: string;
  stateId: string;
  text: string;
  position: number;
}

export interface State {
  id: string;
  boardId: string;
  name: string;
  position: number;
  color: string;
}

export interface Board {
  id: string;
  name: string;
  owner: string;
}
