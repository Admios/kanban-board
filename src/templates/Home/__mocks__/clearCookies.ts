export const clearCookies = jest.fn().mockResolvedValue(undefined);
