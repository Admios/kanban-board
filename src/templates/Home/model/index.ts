import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable, produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";

export type { State, Task } from "./types";

type HomeState = Immutable<{
  tasks: Record<string, Task[]>;
  states: Record<string, State>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTasks: TaskDTO[],
    initialStates: StateDTO[],
    initialUser?: UserDTO,
  ): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates, initialUser) {
    const statesMap: Record<string, State> = {};
    const tasksMap: Record<string, Task[]> = {};

    initialStates.forEach((backendState) => {
      const result: State = {
        ...backendState,
        id: backendState.id,
        position: backendState.position,
      };
      statesMap[result.id] = result;
    });

    initialTasks.forEach((backendTask) => {
      const { stateId } = backendTask;
      if (!tasksMap[stateId]) {
        tasksMap[stateId] = [];
      }
      tasksMap[stateId][backendTask.position] = {
        ...backendTask,
        stateId,
        id: backendTask.id,
        position: backendTask.position,
      };
    });

    set({ tasks: tasksMap, states: statesMap, user: initialUser });
  },

  addTask: (newTask) => {
    const result = { ...newTask };
    const tasks = produce(get().tasks, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }

      draft[newTask.stateId].push(result);
    });

    set({ tasks });
    return result;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const tasks = produce(get().tasks, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task.id !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask);
      } else {
        destinationState.push(newTask);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasks });
  },

  addState: (newState) => {
    const currentModel = get();
    const tasks = produce(currentModel.tasks, (draft) => {
      draft[newState.id] = [];
    });
    const states = produce(currentModel.states, (draft) => {
      const position = Object.keys(currentModel.states).length;
      draft[newState.id] = { ...newState, position };
    });

    set({ states, tasks });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task.id === taskId);
        if (taskIndex > -1) {
          stateTasks[taskIndex] = {
            ...stateTasks[taskIndex],
            ...updatedValues,
          };
          break;
        }
      }
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task?.id === taskId);
        if (taskIndex > -1) {
          stateTasks.splice(taskIndex, 1);
          break;
        }
      }
    });

    set({ tasks });
  },
});

export const useZustand = create(stateCreator);
