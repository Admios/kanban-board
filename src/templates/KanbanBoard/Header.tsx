import { useState } from "react";
import { useRouter } from "next/navigation";
import { useDisclosure, useColorModeValue } from "@chakra-ui/react";
import { HamburgerIcon, CloseIcon, AddIcon } from "@chakra-ui/icons";
import {
  Box,
  Flex,
  Avatar,
  HStack,
  IconButton,
  Button,
  Menu,
  MenuButton,
  MenuList,
  MenuItem,
  MenuDivider,
  Stack,
  Center,
  Image,
} from "@chakra-ui/react";

import { clearCookies } from "./clearCookies";
import { useZustand } from "./model";

interface NavLinkProps {
  children: React.ReactNode;
}

interface HeaderProps {
  handleCreateRandomTasks?: () => void;
  onOpenStateDialog?: () => void;
}

const NavLink = ({ children }: NavLinkProps) => (
  <Box
    as="a"
    px={2}
    py={1}
    rounded="md"
    _hover={{
      textDecoration: "none",
      bg: useColorModeValue("gray.200", "gray.700"),
    }}
    href="#"
  >
    {children}
  </Box>
);

const Links = ["Dashboard"];

export function Header({
  handleCreateRandomTasks,
  onOpenStateDialog,
}: HeaderProps) {
  const user = useZustand((store) => store.user);
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const router = useRouter();

  function handleLogin() {
    router.push("/login");
  }

  async function handleLogout() {
    await clearCookies();
  }

  return (
    <>
      <Box bg={useColorModeValue("gray.100", "gray.900")} px={4}>
        <Flex h={16} alignItems="center" justifyContent="space-between">
          <IconButton
            size="md"
            icon={isOpen ? <CloseIcon /> : <HamburgerIcon />}
            aria-label="Open Menu"
            display={{ md: "none" }}
            onClick={isOpen ? onClose : onOpen}
          />
          <HStack spacing={8} alignItems="center">
            <Box>
              <Image src="admios-logo.svg" alt="" />
            </Box>
            <HStack as="nav" spacing={4} display={{ base: "none", md: "flex" }}>
              {Links.map((link) => (
                <NavLink key={link}>{link}</NavLink>
              ))}
            </HStack>
          </HStack>
          <Flex alignItems="center">
            <Menu
              isOpen={isMenuOpen}
              onClose={() => setIsMenuOpen(!isMenuOpen)}
            >
              <MenuButton
                as={Button}
                variant="solid"
                colorScheme="teal"
                size="sm"
                mr={4}
                leftIcon={<AddIcon />}
                onClick={() => setIsMenuOpen(!isMenuOpen)}
              >
                Action
              </MenuButton>
              <MenuList>
                <MenuItem data-testid="add-state-button" onClick={onOpenStateDialog}>Add State</MenuItem>
                <MenuItem data-testid="create-random-task-button" onClick={handleCreateRandomTasks}>
                  Create 10 random tasks
                </MenuItem>
              </MenuList>
            </Menu>
            <Menu>
              <MenuButton
                as={Button}
                rounded="full"
                variant="link"
                cursor="pointer"
                minW={0}
                data-testid="user-menu-button"
              >
                <Avatar
                  size="sm"
                  bg="blue.500"
                  name={user ? user.username : "guest"}
                  src={user ? "" : "avatar.svg"}
                />
              </MenuButton>
              <MenuList alignItems="center">
                <br />
                <Center>
                  <Avatar
                    size="2xl"
                    bg="blue.500"
                    name={user ? user.username : "guest"}
                    src={user ? "" : "avatar.svg"}
                  />
                </Center>
                <br />
                <Center>
                  <p>{user?.username ? user.username : "guest"} </p>
                </Center>
                <br />
                <MenuDivider />
                {user ? (
                  <MenuItem data-testid="logout-button" onClick={handleLogout}>Logout</MenuItem>
                ) : (
                  <MenuItem data-testid="login-button" onClick={handleLogin}>Login / Register</MenuItem>
                )}
              </MenuList>
            </Menu>
          </Flex>
        </Flex>

        {isOpen ? (
          <Box pb={4} display={{ md: "none" }}>
            <Stack as="nav" spacing={4}>
              {Links.map((link) => (
                <NavLink key={link}>{link}</NavLink>
              ))}
            </Stack>
          </Box>
        ) : null}
      </Box>
    </>
  );
}
