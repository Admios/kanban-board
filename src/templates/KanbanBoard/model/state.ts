import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
