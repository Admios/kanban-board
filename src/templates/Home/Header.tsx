import { Box, Button, Tag } from "@chakra-ui/react";
import { useRouter } from "next/navigation";
import { clearCookies } from "./clearCookies";
import { useZustand } from "./state";

function UserButton() {
  const user = useZustand((store) => store.user);
  const router = useRouter();

  function handleLogin() {
    router.push("/login");
  }

  async function handleLogout() {
    await clearCookies();
    router.push("/login");
  }

  if (!user) {
    return (
      <Button colorScheme="blue" onClick={handleLogin}>
        Login / Register
      </Button>
    );
  }

  return (
    <>
      <Tag colorScheme="blue">Your Username is: {user.username}</Tag>
      <Button colorScheme="blue" onClick={handleLogout}>
        Logout
      </Button>
    </>
  );
}

interface HeaderProps {
  handleCreateRandomTasks: () => void;
  onOpenColumnDialog: () => void;
}

export function Header({
  handleCreateRandomTasks,
  onOpenColumnDialog,
}: HeaderProps) {
  return (
    <Box as="header" display="flex" gap="4">
      <UserButton />
      <Button colorScheme="blue" onClick={onOpenColumnDialog}>
        Add Column
      </Button>
      <Button colorScheme="blue" onClick={handleCreateRandomTasks}>
        Create 10 random tasks
      </Button>
    </Box>
  );
}
