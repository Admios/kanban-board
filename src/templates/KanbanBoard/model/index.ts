import { produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";
import { KanbanState, KanbanActions } from "./state";

export type { State, Task } from "./types";

const stateCreator: StateCreator<KanbanState & KanbanActions> = (set, get) => ({
  statesOrder: [],
  tasksOrder: {},
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates) {
    const statesOrderSparseArray: string[] = [];
    const tasksOrder: Record<string, string[]> = {};
    const states: Record<string, State> = {};
    const tasks: Record<string, Task> = {};

    initialTasks.forEach((backendTask) => {
      tasks[backendTask.id] = backendTask;
      if (!tasksOrder[backendTask.stateId]) {
        tasksOrder[backendTask.stateId] = [];
      }
      // Warning: saved as sparse array! Compression is done later
      tasksOrder[backendTask.stateId][backendTask.position] = backendTask.id;
    });

    initialStates.forEach((backendState) => {
      states[backendState.id] = backendState;
      // Warning: saved as sparse array! compressed later
      statesOrderSparseArray[backendState.position] = backendState.id;
      // Compress sparse array
      tasksOrder[backendState.id] = (tasksOrder[backendState.id] ?? []).filter(
        (id) => !!id,
      );
    });

    // Compress sparse array
    const statesOrder = statesOrderSparseArray.filter((id) => !!id);

    set({
      statesOrder,
      tasksOrder,
      tasks,
      states,
    });
  },

  setUser(user) {
    set({ user });
  },

  addTask: (newTask) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[newTask.stateId] = newTask;
    });

    const tasksOrder = produce(get().tasksOrder, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }
      draft[newTask.stateId].push(newTask.id);
    });

    set({ tasks, tasksOrder });
    return newTask;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const newTasksOrder = produce(get().tasksOrder, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask.id);
      } else {
        destinationState.push(newTask.id);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasksOrder: newTasksOrder });
  },

  addState: (newState) => {
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[newState.id] = [];
    });

    const statesOrder = produce(get().statesOrder, (draft) => {
      draft.push(newState.id);
    });

    const states = produce(get().states, (draft) => {
      draft[newState.id] = newState;
    });

    set({ states, tasksOrder, statesOrder });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[taskId] = { ...draft[taskId], ...updatedValues };
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const { stateId } = get().tasks[taskId];
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[stateId] = draft[stateId].filter((id) => id !== taskId);
    });

    const tasks = produce(get().tasks, (draft) => {
      delete draft[taskId];
    });

    set({ tasks, tasksOrder });
  },
});

export const useZustand = create(stateCreator);
