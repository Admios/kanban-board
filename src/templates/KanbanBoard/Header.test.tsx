import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useRouter } from "next/navigation";
import { Header } from "./Header";
import { clearCookies } from "./clearCookies";
import { useZustand } from "./model";

jest.mock("./clearCookies.ts");
jest.mock("./kanbanActions.ts");

it("should launch login when button is pressed", async () => {
  const { push } = (useRouter as jest.Mock)();
  const { container } = render(<Header />);
  expect(container).toMatchSnapshot("Default Header");

  // Click the button
  const button = await screen.findByTestId("login-button");
  expect(button).toBeInTheDocument();

  await userEvent.click(button);
  expect(push).toHaveBeenCalledWith("/login");
});

it("should launch logout when button is pressed", async () => {
  useZustand.setState({
    user: {
      email: "test",
    },
  });

  const { push } = (useRouter as jest.Mock)();
  const { container } = render(<Header />);
  expect(container).toMatchSnapshot("Logged in Header");

  // Click the button
  const button = await screen.findByTestId("logout-button");
  expect(button).toBeInTheDocument();

  await userEvent.click(button);
  expect(clearCookies).toHaveBeenCalled();
  expect(push).toHaveBeenCalledWith("/login");

  useZustand.setState({});
});
