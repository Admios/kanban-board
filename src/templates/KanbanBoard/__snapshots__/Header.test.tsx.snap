// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should launch login when button is pressed: Default Header 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1m65r4j"
    >
      <div
        class="css-q4hjtx"
      >
        <button
          aria-label="Open Menu"
          class="chakra-button css-1s5iq1h"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="chakra-stack css-chgbhg"
        >
          <div
            class="css-0"
          >
            <img
              alt=""
              class="chakra-image css-0"
              src="admios-logo.svg"
            />
          </div>
          <nav
            class="chakra-stack css-1eq794"
          >
            <a
              class="css-1hol3b3"
              href="/"
            >
              Board
            </a>
          </nav>
        </div>
        <div
          class="css-70qvj9"
        >
          <button
            aria-controls="menu-list-:r3:"
            aria-expanded="false"
            aria-haspopup="menu"
            class="chakra-button chakra-menu__menu-button css-czglya"
            id="menu-button-:r3:"
            type="button"
          >
            <span
              class="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="css-xl71ch"
            >
              Action
            </span>
          </button>
          <div
            class="css-0"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <div
              aria-orientation="vertical"
              class="chakra-menu__menu-list css-sjt5nk"
              id="menu-list-:r3:"
              role="menu"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
              tabindex="-1"
            >
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="0"
                data-testid="add-state-button"
                id="menu-list-:r3:-menuitem-:r4:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Add State
              </button>
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="1"
                data-testid="create-random-task-button"
                id="menu-list-:r3:-menuitem-:r5:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Create 10 random tasks
              </button>
            </div>
          </div>
          <button
            aria-controls="menu-list-:r7:"
            aria-expanded="false"
            aria-haspopup="menu"
            class="chakra-button chakra-menu__menu-button css-aqz4t6"
            data-testid="user-menu-button"
            id="menu-button-:r7:"
            type="button"
          >
            <span
              class="css-xl71ch"
            >
              <span
                class="chakra-avatar css-332l87"
              >
                <div
                  aria-label="guest"
                  class="chakra-avatar__initials css-0"
                  role="img"
                >
                  g
                </div>
              </span>
            </span>
          </button>
          <div
            class="css-0"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <div
              aria-orientation="vertical"
              class="chakra-menu__menu-list css-1jabnse"
              id="menu-list-:r7:"
              role="menu"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
              tabindex="-1"
            >
              <br />
              <div
                class="css-gmuwbf"
              >
                <span
                  class="chakra-avatar css-332l87"
                >
                  <div
                    aria-label="guest"
                    class="chakra-avatar__initials css-0"
                    role="img"
                  >
                    g
                  </div>
                </span>
              </div>
              <br />
              <div
                class="css-gmuwbf"
              >
                <p>
                  guest
                   
                </p>
              </div>
              <br />
              <hr
                aria-orientation="horizontal"
                class="chakra-menu__divider css-0"
              />
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="0"
                data-testid="login-button"
                id="menu-list-:r7:-menuitem-:r8:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Login / Register
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1kxrbkv"
    >
      <div
        class="css-erspa2"
      />
    </div>
  </div>
</div>
`;

exports[`should launch logout when button is pressed: Logged in Header 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1m65r4j"
    >
      <div
        class="css-q4hjtx"
      >
        <button
          aria-label="Open Menu"
          class="chakra-button css-1s5iq1h"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="chakra-stack css-chgbhg"
        >
          <div
            class="css-0"
          >
            <img
              alt=""
              class="chakra-image css-0"
              src="admios-logo.svg"
            />
          </div>
          <nav
            class="chakra-stack css-1eq794"
          >
            <a
              class="css-1hol3b3"
              href="/"
            >
              Board
            </a>
          </nav>
        </div>
        <div
          class="css-70qvj9"
        >
          <button
            aria-controls="menu-list-:rf:"
            aria-expanded="false"
            aria-haspopup="menu"
            class="chakra-button chakra-menu__menu-button css-czglya"
            id="menu-button-:rf:"
            type="button"
          >
            <span
              class="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="css-xl71ch"
            >
              Action
            </span>
          </button>
          <div
            class="css-0"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <div
              aria-orientation="vertical"
              class="chakra-menu__menu-list css-sjt5nk"
              id="menu-list-:rf:"
              role="menu"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
              tabindex="-1"
            >
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="0"
                data-testid="add-state-button"
                id="menu-list-:rf:-menuitem-:rg:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Add State
              </button>
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="1"
                data-testid="create-random-task-button"
                id="menu-list-:rf:-menuitem-:rh:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Create 10 random tasks
              </button>
            </div>
          </div>
          <button
            aria-controls="menu-list-:rj:"
            aria-expanded="false"
            aria-haspopup="menu"
            class="chakra-button chakra-menu__menu-button css-aqz4t6"
            data-testid="user-menu-button"
            id="menu-button-:rj:"
            type="button"
          >
            <span
              class="css-xl71ch"
            >
              <span
                class="chakra-avatar css-332l87"
              >
                <div
                  aria-label="test"
                  class="chakra-avatar__initials css-0"
                  role="img"
                >
                  t
                </div>
              </span>
            </span>
          </button>
          <div
            class="css-0"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <div
              aria-orientation="vertical"
              class="chakra-menu__menu-list css-1jabnse"
              id="menu-list-:rj:"
              role="menu"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
              tabindex="-1"
            >
              <br />
              <div
                class="css-gmuwbf"
              >
                <span
                  class="chakra-avatar css-332l87"
                >
                  <div
                    aria-label="test"
                    class="chakra-avatar__initials css-0"
                    role="img"
                  >
                    t
                  </div>
                </span>
              </div>
              <br />
              <div
                class="css-gmuwbf"
              >
                <p>
                  test
                   
                </p>
              </div>
              <br />
              <hr
                aria-orientation="horizontal"
                class="chakra-menu__divider css-0"
              />
              <button
                class="chakra-menu__menuitem css-1k756t9"
                data-index="0"
                data-testid="logout-button"
                id="menu-list-:rj:-menuitem-:rk:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Logout
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1kxrbkv"
    >
      <div
        class="css-erspa2"
      />
    </div>
  </div>
</div>
`;
